### PR TITLE
Pools/per host config

### PR DESCRIPTION
**Pull request by NelsonVides**
What I wanted to have was to allow to have a default global pool, and another default for a single_host_type for example. But the current TOML config doesn’t let me have two pools tagged default. Allow for this.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/210355) / elasticsearch_and_cassandra_mnesia / a964a4126efe8af2e111714273a7aa46a3fe4ed5
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4235/210355/elasticsearch_and_cassandra_mnesia.26.1.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4235/210355/elasticsearch_and_cassandra_mnesia.26.1.2-amd64/big/ct_run.test%40fcb9e44746e1.2024-03-05_14.39.43/index.html)
**OK: 429** / Failed: 0 / **User-skipped: 43** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/210492) / elasticsearch_and_cassandra_mnesia / 80acbf111c6d320dbe40174333d92014531b7db4
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4235/210492/elasticsearch_and_cassandra_mnesia.26.1.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4235/210492/elasticsearch_and_cassandra_mnesia.26.1.2-amd64/big/ct_run.test%40b0d9eea9d8f1.2024-03-06_08.57.25/index.html)
**OK: 429** / Failed: 0 / **User-skipped: 43** / Auto-skipped: 0

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/210625) / elasticsearch_and_cassandra_mnesia / 075d95750f070d51a00e3fdf4be4e7952da0137b
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4235/210625/elasticsearch_and_cassandra_mnesia.26.1.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4235/210625/elasticsearch_and_cassandra_mnesia.26.1.2-amd64/big/ct_run.test%40b4a46f1dbca6.2024-03-06_10.03.01/index.html)
**OK: 429** / Failed: 0 / **User-skipped: 43** / Auto-skipped: 0

File: src/config/mongoose_config_spec.erl [hunk 1]
```diff
@@ -474,11 +489,14 @@ wpool(ExtraDefaults) ->
                                      <<"strategy">> => best_worker,
                                      <<"call_timeout">> => 5000}, ExtraDefaults)}.
 
-outgoing_pool_extra(Type) ->
+outgoing_pool_extra(host_config, Type) ->
+    #section{items = #{<<"connection">> => outgoing_pool_connection(Type)},
+             process = fun ?MODULE:process_host_config_pool/2
+            };
+outgoing_pool_extra(global_config, Type) ->
+    Scopes = [global, host_type, host], %% TODO deprecated
```
Comment (chrzaszcz) [inline src/config/mongoose_config_spec.erl:497]:
```suggestion
    Scopes = [global, host_type, host], %% TODO 'host' is deprecated
```

File: src/config/mongoose_config_spec.erl [hunk 2]
```diff
@@ -474,11 +489,14 @@ wpool(ExtraDefaults) ->
...
     #section{items = #{<<"scope">> => #option{type = atom,
-                                              validate = {enum, [global,
-                                                                 host_type, single_host_type,
-                                                                 host, single_host]}}, %% TODO deprecated
+                                              validate = {enum, Scopes}},
                        <<"host_type">> => #option{type = binary,
                                                   validate = non_empty},
                        <<"host">> => #option{type = binary, %% TODO deprecated
```
Comment (chrzaszcz) [inline src/config/mongoose_config_spec.erl:502]: Shouldn't these be removed now?

File: src/config/mongoose_config_spec.erl [hunk 3]
```diff
@@ -1081,7 +1099,7 @@ check_auth_method(Method, Opts) ->
         false -> error(#{what => missing_section_for_auth_method, auth_method => Method})
     end.
 
-process_pool([Tag, Type|_], AllOpts = #{scope := ScopeIn, connection := Connection}) ->
+process_pool([Tag, Type | _], AllOpts = #{scope := ScopeIn, connection := Connection}) ->
     Scope = pool_scope(ScopeIn, maps:get(host_type, AllOpts, maps:get(host, AllOpts, none))),
```
Comment (chrzaszcz) [inline src/config/mongoose_config_spec.erl:1103]:
```suggestion
    Scope = pool_scope(ScopeIn),
```

File: src/wpool/mongoose_wpool.erl [hunk 4]
```diff
@@ -121,8 +113,12 @@ start_configured_pools(PoolsIn) ->
     start_configured_pools(PoolsIn, ?ALL_HOST_TYPES).
 
 start_configured_pools(PoolsIn, HostTypes) ->
+    HostTypeSpecific = get_host_type_specific_pools(HostTypes),
+    start_configured_pools(PoolsIn, HostTypeSpecific, HostTypes).
+
+start_configured_pools(PoolsIn, HostTypeSpecific, HostTypes) ->
+    Pools = expand_pools(PoolsIn, HostTypeSpecific, HostTypes),
     [call_callback(init, PoolType, []) || PoolType <- get_unique_types(PoolsIn)],
```
Comment (chrzaszcz) [inline src/wpool/mongoose_wpool.erl:121]: Why is init not called for host-type-specific pools? They don't need to appear on the top level at all.

**Review by chrzaszcz — COMMENTED** [review]
It's going in the right direction. I added some comments.

File: src/config/mongoose_config_spec.erl (hunk 2 above)
Comment (NelsonVides) [reply src/config/mongoose_config_spec.erl:502]: Truuuu 🤔

**Comment by mongoose-im** [comment]
[elasticsearch_and_cassandra_26](https://circleci.com/gh/esl/MongooseIM/210889) / elasticsearch_and_cassandra_mnesia / daf4c017adfd2f7e90a08af7f1c745575c208d99
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/4235/210889/elasticsearch_and_cassandra_mnesia.26.1.2-amd64)/ [big](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/4235/210889/elasticsearch_and_cassandra_mnesia.26.1.2-amd64/big/ct_run.test%4014066b889722.2024-03-07_14.44.52/index.html)
**OK: 429** / Failed: 0 / **User-skipped: 43** / Auto-skipped: 0

**Review by chrzaszcz — APPROVED** [review]
Looks good 👍

**Pull request merged**